lib/src/auth/auth.dart

```diff
@@ -0,0 +1,98 @@
+/// Source:  https://github.com/mongo-dart/mongo_dart/blob/c761839efbf47ec556f853dec85debb4cb9370f7/lib/src/auth/auth.dart
+
+// (The MIT License)
+//
+// Copyright (c) 2012 Vadim Tsushko (vadimtsushko@gmail.com)
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// 'Software'), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import 'dart:math';
+
+import 'package:postgres/src/auth/sasl/scram_sha256_authenticator.dart';
+
+import '../../postgres.dart';
+import '../server_messages.dart';
+import 'md5/md5_authenticator.dart';
+
+enum AuthenticationScheme { MD5, SCRAM_SHA_256 }
+
+abstract class Authenticator {
+  static String? name;
+  late final PostgreSQLConnection connection;
+
+  Authenticator(this.connection);
+
+  void init();
+
+  void onMessage(AuthenticationMessage message);
+}
+
+Authenticator createAuthenticator(PostgreSQLConnection connection, UsernamePasswordCredential credentials) {
+  switch (connection.authenticationScheme) {
+    case AuthenticationScheme.MD5:
+      return MD5Authenticator(connection, credentials);
+    case AuthenticationScheme.SCRAM_SHA_256:
+      return ScramSha256Authenticator(connection, credentials);
+    default:
+      throw PostgreSQLException("Authenticator wasn't specified");
+  }
+}
+
+class UsernamePasswordCredential {
+  String? username;
+  String? password; // TODO: Encrypt this to secureString
+}
+
+abstract class RandomStringGenerator {
+  static const String allowedCharacters = '!"#\'\$%&()*+-./0123456789:;<=>?@'
+  'ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~';
+  String generate(int length);
+}
+
+class CryptoStrengthStringGenerator extends RandomStringGenerator {
+  @override
+  String generate(int length) {
+    final random = Random.secure();
+    final allowedCodeUnits = RandomStringGenerator.allowedCharacters.codeUnits;
+
+    final max = allowedCodeUnits.length - 1;
```

Comment:
This is a strange choice, as it will never select the last character from the allowed character list. (random.nextInt is >=0 and <max).

Reply:
Sounds reasonable, this was introduced in https://github.com/mongo-dart/mongo_dart/pull/86
Maybe @Pacane has an idea :)

Reply:
I think I thought the upper bound was inclusive at that time. Or could just be a bug. It was years ago lol.

Reply:
This is fixed in the sasl_scram library now :)